ssl/quic/quic_channel.c

```diff
@@ -0,0 +1,1639 @@
+                                                   0))
+        goto err;
+
+    if (!ossl_quic_wire_encode_transport_param_int(&wpkt, QUIC_TPARAM_INITIAL_MAX_STREAMS_UNI,
+                                                   0))
+        goto err;
+
+    if (!WPACKET_get_total_written(&wpkt, &buf_len))
+        goto err;
+
+    ch->local_transport_params = (unsigned char *)buf_mem->data;
+    buf_mem->data = NULL;
+
+    if (!WPACKET_finish(&wpkt))
+        goto err;
+
+    wpkt_valid = 0;
+
+    if (!ossl_quic_dhs_set_transport_params(ch->dhs, ch->local_transport_params,
+                                            buf_len))
+        goto err;
+
+    ok = 1;
+err:
+    if (wpkt_valid)
+        WPACKET_cleanup(&wpkt);
+    BUF_MEM_free(buf_mem);
+    return ok;
+}
+
+/*
+ * QUIC Channel: Ticker-Mutator
+ * ============================
+ */
+
+/*
+ * The central ticker function called by the reactor. This does everything, or
+ * at least everything network I/O related. Best effort - not allowed to fail
+ * "loudly".
+ */
+static void ch_tick(QUIC_TICK_RESULT *res, void *arg)
+{
+    OSSL_TIME now, deadline;
+    QUIC_CHANNEL *ch = arg;
+
+    /*
+     * When we tick the QUIC connection, we do everything we need to do
+     * periodically. In order, we:
+     *
+     *   - handle any incoming data from the network;
+     *   - handle any timer events which are due to fire (ACKM, etc.)
+     *   - write any data to the network due to be sent, to the extent
+     *     possible;
+     *   - determine the time at which we should next be ticked.
+     */
+
+    /* If we are in the TERMINATED state, there is nothing to do. */
+    if (ossl_quic_channel_is_terminated(ch)) {
+        res->want_net_read  = 0;
+        res->want_net_write = 0;
+        res->tick_deadline  = ossl_time_infinite();
+        return;
+    }
+
+    /*
+     * If we are in the TERMINATING state, check if the terminating timer has
+     * expired.
+     */
+    if (ossl_quic_channel_is_terminating(ch)) {
+        now = ossl_time_now();
+
+        if (ossl_time_compare(now, ch->terminate_deadline) >= 0) {
+            ch_on_terminating_timeout(ch);
+            res->want_net_read  = 0;
+            res->want_net_write = 0;
+            res->tick_deadline  = ossl_time_infinite();
+            return; /* abort normal processing, nothing to do */
+        }
+    }
+
+    do {
+        /* Handle any incoming data from the network. */
+        ch_rx(ch);
+
+        /*
+         * Allow the handshake layer to check for any new incoming data and generate
+         * new outgoing data.
+         */
+        ch->have_new_secret = 0;
+        ossl_quic_dhs_tick(ch->dhs);
+
+        /*
+         * If the handshake layer gave us a new secret, we need to do RX again
+         * because packets that were not previously processable and were
+         * deferred might now be processable.
+         */
+    } while (ch->have_new_secret);
```

Comment:
I wonder whether we could avoid this loop altogether and instead immediately process deferred packets in the yield_secret_cb? The advantage of doing that is that we don't have to exit where we are in the TLS stack, i.e. with this the way things are the TLS stack will attempt to read the EncryptedExtensions message. It will always fail first time and exit all the way back up the stack to here, only to then re-enter the TLS stack via the tls_tick function to go all the way back into the state it was in previously to attempt the read again. This strikes me as sub-optimal and we could avoid it by processing the deferred packets in yield_secret_cb.

Reply:
Maybe add `TODO(QUIC)` somewhere?